### ISSUE TITLE
Denote <-> Logseq migration

### ISSUE THREAD

**Issue by hapst3r** [bitspook/notes-migrator#1]
Hej fellow,

I wrote in your annoucement thread on reddit that I had been thinking about writing a some elisp code to transition denote notes to logseq for work purposes. Again, I appreciate your idea to join forces :)

DISCLAIMER: I am a newbie when it comes to programming. Even though I've been learning on my own for quite a while, I might make stupid mistakes and ask stupid questions.

With that out of the way, here are my thoughts on a Denote -> Logseq migration after taking a look at your code.  Because I don't want to be overwhelmed, I will not think about the other direction for the moment.

### Following links
Contrary to what my initial testing suggested, the migration from Denote to Logseq is greatly facilitated by the fact that Logseq features some behaviours I thought I'd have to manually implement. I can easily just import my denote folder and open them into the program no problem. Most importantly, it does read and correctly formats the in-buffer settings. 

The only problem with Denote -> Logseq is that of links between notes. When I link files, I always use `org-insert-link`, using the denote: type for autocompletion. The upside is that even when files are renamed, the link persists because it uses the IDENTIFIER generated when creating the note. Logseq, in contrast, uses the file's TITLE to identify a note.

### Proposed implementation

1. Copy all contents of DENOTE-DIRECTORY to LOGSEQ-DIRECTORY. I believe it makes sense to define global variables for this.
2. List the contents of the DENOTE-DIRECTORY.
3. For each file, check for links. For each link, find according file from directory contents (2) using the IDENTIFIER. Then, replace [[denote:20221003T11:1503][my description]] with [[TITLE]]. Save.

### What's left to do
Obviously, the migration logseq -> denote. This is further complicated by the fact that while note creation in denote kind of enforces the receipt of information such as IDENTIFIER, FILETAGS and TITLE and consequently puts them on top of the file, logseq by default does not do this.

I will have to do more research on this, but for the moment I believe it is easiest to first concentrate on Denote -> Logseq and only later care about the other way around.

What do you think?

**Comment by bitspook** [comment]
I have created a [new branch](https://github.com/bitspook/notes-migrator/tree/1-denote-logseq-migration) for this issue, and have sent you an invitation so you can push to this repo directly. You can use that, or you can work in a separate fork and create a PR when you think it's good.

I'd suggest creating a separate `.el` file for this migrator; we'll rename `notes-migrator.el` to `org-roam-to-denote.el` or something when a new migrator is ready.

---

Your overall approach sounds good. denote->logseq and logseq<-denote seem orthogonal, i.e looks like they can be done independently and don't really depend on each other. So it'd be a good idea to focus on denote->logseq first. For logseq<-denote, we need to figure out a way to uniquely identify files in logseq with something other than the file's name.

For converting denote links to logseq, you can reuse [nm--convert-roam-links-to-denote](https://github.com/bitspook/notes-migrator/blob/4c6495adff8f1b754a3d1909c2d5a6082f133960/notes-migrator.el#L66-L84) to some degree. It'll give you some idea of using [org-element api](https://orgmode.org/worg/dev/org-element-api.html) and manipulating org files in general.

---

>  I am a newbie when it comes to programming. Even though I've been learning on my own for quite a while, I might make stupid mistakes and ask stupid questions.

You're most welcome to ask/share whatever you feel like. You can also join [Emacs matrix room](https://app.element.io/#/room/#emacs:matrix.org); questions asked there get answered lot more quickly and by wiser folks.

And don't worry about coming up with a clean solution or clean code; we can polish it up together. Feel free to just copy the bits you need from `notes-migrator.el` or elsewhere and just drop it in your code. It'd be helpful if you add references (in comments) though, they'll help with trying to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) things up in a future iteration when clean code becomes a priority.

PS feel free to hit me up any time on twitter/email/telegram (same handle) if you want to have a quick chat or bounce some ideas around.

**Comment by hapst3r** [comment]
hej man,

I appreciate your quick response.

I just cloned the the repo and copied the contents on notes-migrator.el to denote-to-logseq.el, so I have a starting point in the new branch.

I am new to working with others in a git repo, so feel free to tell me when my usage of git is wrong and/or not idiomatic.

Thanks for the pointers, I will start working on the files starting from tomorrow I think.

**EDIT:** Hej ho, I've got a first question. In `nm--convert-roam-links-to-denote`, there is an &optional element filename, which, according to the docstring, is supposed to help with debugging. But then, I can't find any reference to said filename in the function itself.

Since I am struggling to wrap my head around the workings of ORG-ELEMENT theoretically speaking (without something tangible), I'd love to use a FILENAME argument so that I can experiment. I guess I would use `WITH-TEMP-BUFFER` then, right?
